File: backend/main/java/de/griefed/serverpackcreator/utilities/versionchecker/VersionChecker.java
/* Copyright (C) 2022  Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.utilities.versionchecker;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.griefed.serverpackcreator.i18n.LocalizationManager;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.web.client.HttpClientErrorException;

import java.util.ArrayList;
import java.util.List;

/**
 * Baseclass from wich GitHub and GitLab checks extend. This class mainly provides the logic for comparing versions against
 * each other to find out which is newer.
 * @author Griefed
 */
public abstract class VersionChecker {

    private static final Logger LOG = LogManager.getLogger(VersionChecker.class);

    private List<String> allVersions;
    protected LocalizationManager LOCALIZATIONMANAGER;

    /**
     * Check whether an update/newer version is available for the given version. If you want to check for PreReleases, too,
     * then make sure to pass <code>true</code> for <code>checkForPreReleases</code>.
     * @author Griefed
     * @param currentVersion String. The current version of the app.
     * @param checkForPreReleases Boolean. <code>false</code> if you do not want to check for PreReleases. <code>true</code>
     *                            if you want to check for PreReleases as well.
     * @return String. Returns a concatenated String whether an update is available or not. Examples:<br>
     * No update available: <code>No updates available.</code><br>
     * New release available: <code>Current version: 2.0.0. A new release is available: 2.1.1. Download available at: https://github.com/Griefed/ServerPackCreator/releases/tag/2.1.1</code>
     * New prerelease available: <code>Current version: 2.0.0. A new PreRelease is available: 3.0.0-alpha.14. Download available at: https://github.com/Griefed/ServerPackCreator/releases/tag/3.0.0-alpha.14</code>
     */
    public String checkForUpdate(String currentVersion, boolean checkForPreReleases) {

        LOG.debug("Current version: " + currentVersion);

        try {

            String newVersion = isUpdateAvailable(currentVersion, checkForPreReleases);

            if (newVersion.equals("up_to_date")) {
                return LOCALIZATIONMANAGER.getLocalizedString("updates.log.info.none");
            } else {
                return newVersion + ";" + getDownloadUrl(newVersion);
            }

        } catch (NumberFormatException ex) {
            LOG.error("A version could not be parsed into integers.", ex);
            return LOCALIZATIONMANAGER.getLocalizedString("updates.log.info.none");
        }
    }

    /**
     * Check for new versions in beta, alpha and regular release channels. If <code>checkForPreRelease</code> is false,
     * only regular releases are checked.
     * @author Griefed
     * @param currentVersion String. The current version of the app.
     * @param checkForPreReleases Boolean. <code>false</code> if you do not want to check for PreReleases. <code>true</code>
     *                            if you want to check for PreReleases as well.
     * @return String. Returns the available update version. If no update is available, then <code>up_to_date</code> is returned.
     */
    private String isUpdateAvailable(String currentVersion, boolean checkForPreReleases) {

        if (isNewBetaAvailable(currentVersion, checkForPreReleases) && checkForPreReleases) return latestBeta();

        if (isNewAlphaAvailable(currentVersion, checkForPreReleases) && checkForPreReleases) return latestAlpha();

        if (isNewSemanticVersion(currentVersion, latestVersion(checkForPreReleases))) return latestVersion(checkForPreReleases);

        return "up_to_date";
    }

    /**
     * Check the given current version against the given new version whether the new version is actually newer.
     * Checks are performed with the semantic release-formatting, e.g. 1.2.3, 2.3.4, 6.6.6
     * @author Griefed
     * @param currentVersion String. Current version to check against <code>newVersion</code>.
     * @param newVersion String. New version to check against <code>currentVersion</code>.
     * @return Boolean. Returns <code>true</code> if the new version is indeed newer than the current version. Otherwise
     * <code>false</code>.
     * @throws NumberFormatException Thrown if the passed <code>currentVersion</code> or <code>newVersion</code> can not be
     * parsed into integers.
     */
    private boolean isNewSemanticVersion(String currentVersion, String newVersion) throws NumberFormatException {

        LOG.debug("Current version: " + currentVersion);
        LOG.debug("New version: " + newVersion);

        int newMajor,newMinor,newPatch,currentMajor,currentMinor,currentPatch;

        newMajor = Integer.parseInt(newVersion.substring(0,1));
        newMinor = Integer.parseInt(newVersion.substring(2,3));
        newPatch = Integer.parseInt(newVersion.substring(4,5));

        currentMajor = Integer.parseInt(currentVersion.substring(0,1));
        currentMinor = Integer.parseInt(currentVersion.substring(2,3));
        currentPatch = Integer.parseInt(currentVersion.substring(4,5));

        if (newMajor > currentMajor && newMinor >= currentMinor && newPatch >= currentPatch) {
            // new major update
            return true;

        } else if (newMajor >= currentMajor && newMinor > currentMinor && newPatch >= currentPatch) {
            // new minor update
            return true;

            // new patch update if true
        } else
            return newMajor >= currentMajor && newMinor >= currentMinor && newPatch > currentPatch;
    }

    /**
     * Check the given current version against the given new version whether the new version is actually newer or the same.
     * Checks are performed with the semantic release-formatting, e.g. 1.2.3, 2.3.4, 6.6.6
     * @author Griefed
     * @param currentVersion String. Current version to check against <code>newVersion</code>.
     * @param newVersion String. New version to check against <code>currentVersion</code>.
     * @return Boolean. Returns <code>true</code> if the new version is newer than or equal to the current version. Otherwise
     * <code>false</code>.
     * @throws NumberFormatException Thrown if the passed <code>currentVersion</code> or <code>newVersion</code> can not be
     * parsed into integers.
     */
    private boolean isNewOrSameSemanticVersion(String currentVersion, String newVersion) throws NumberFormatException {
        if (Integer.parseInt(newVersion.substring(0,1)) >= Integer.parseInt(currentVersion.substring(0,1))) {
            return true;
        } else if (Integer.parseInt(newVersion.substring(2,3)) >= Integer.parseInt(currentVersion.substring(2,3))) {
            return true;
        } else return Integer.parseInt(newVersion.substring(4,5)) >= Integer.parseInt(currentVersion.substring(4,5));
    }

    /**
     * Check for a new alpha version.
     * @author Griefed
     * @param currentVersion String. The current version to check against available alpha versions.
     * @param checkForPreRelease Boolean. Whether to check for PreReleases.
     * @return Boolean. Returns true if a new alpha release is found.
     * @throws NumberFormatException Thrown if the passed <code>currentVersion</code> can not be
     * parsed into integers.
     */
    private boolean isNewAlphaAvailable(String currentVersion, boolean checkForPreRelease) throws NumberFormatException {

        /*
         * If the current version does not contain alpha and checkForPreRelease is false, do not check for a new alpha
         * release.
         */
        if (!currentVersion.contains("alpha") && !checkForPreRelease) return false;

        // If no alpha releases are available, do not check for new alpha release.
        if (latestAlpha().equals("no_alphas")) return false;

        String latestAlpha = latestAlpha();

        // Check if the given version is older than the latest alpha version by checking semantically. (1.2.3, 2.3.4, 6.6.6)
        if (isNewSemanticVersion(currentVersion, latestAlpha)) return true;

        // If a new alpha, say alpha.5 for the given, say alpha.1, is available, return true.
        return Integer.parseInt(latestAlpha.substring(12)) > Integer.parseInt(currentVersion.substring(12));

    }

    /**
     * Check for a new beta version.
     * @author Griefed
     * @param currentVersion String. The current version to check against available beta versions.
     * @param checkForPreRelease Boolean. Whether to check for PreReleases.
     * @return Boolean. Returns true if a new beta release is found.
     * @throws NumberFormatException Thrown if the passed <code>currentVersion</code> can not be
     * parsed into integers.
     */
    private boolean isNewBetaAvailable(String currentVersion, boolean checkForPreRelease) throws NumberFormatException {

        /*
         * If the current version does not contain beta and checkForPreRelease is false, do not check for a new beta
         * release.
         */
        if (!currentVersion.contains("beta") && !checkForPreRelease) return false;

        // If no beta releases are available, do not check for new beta release.
        if (latestBeta().equals("no_betas")) return false;

        String latestBeta = latestBeta();

        // Check if the given version is older than the latest beta version by checking semantically. (1.2.3, 2.3.4, 6.6.6)
        if (isNewSemanticVersion(currentVersion, latestBeta)) return true;

        // If a new beta, say beta.5 for the given, say beta.1, is available, return true.
        return Integer.parseInt(latestBeta.substring(11)) > Integer.parseInt(currentVersion.substring(11));

    }

    /**
     * Gather all beta versions in a list.
     * @author Griefed
     * @return List String. Returns a list of all available beta versions. If no beta releases are available for the given
     * repository, <code>null</code> is returned.
     */
    private List<String> allBetaVersions() {

        List<String> betaVersions = new ArrayList<>(1000);

        // Check all available versions whether they contain beta and add them to a list of beta versions.
        if (allVersions != null) {
            for (String version : allVersions) {
                if (version.contains("beta") && !betaVersions.contains(version)) {
                    betaVersions.add(version);
                }
            }
        }

        if (betaVersions.size() == 0) {
            return null;
        } else {
            return betaVersions;
        }
    }

    /**
     * Gather all alpha versions in a list.
     * @author Griefed
     * @return List String. Returns a list of all available alpha versions. If no alpha releases are available for the given
     * repository, <code>null</code> is returned.
     */
    private List<String> allAlphaVersions() {

        List<String> alphaVersions = new ArrayList<>(1000);

        // Check all available versions whether they contain alpha and add them to a list of alpha versions.
        if (allVersions != null) {
            for (String version : allVersions) {
                if (version.contains("alpha") && !alphaVersions.contains(version)) {
                    alphaVersions.add(version);
                }
            }
        }

        if (alphaVersions.size() == 0) {
            return null;
        } else {
            return alphaVersions;
        }
    }

    protected abstract List<String> allVersions();

    /**
     * Get the latest beta release.
     * @author Griefed
     * @return String. Returns the latest beta release. If no beta release is available, <code>no_betas</code> is returned.
     * @throws NumberFormatException Thrown if a version can not be parsed into integers.
     */
    protected String latestBeta() throws NumberFormatException {

        List<String> betaVersions = allBetaVersions();
        String beta = "no_betas";

        if (betaVersions != null) {

            beta = betaVersions.get(0);

            for (String betaVersion : betaVersions) {

                if (isNewOrSameSemanticVersion(beta, betaVersion) && Integer.parseInt(betaVersion.substring(11)) >= Integer.parseInt(beta.substring(11))) {
                    beta = betaVersion;
                }
            }

        }

        LOG.debug("Latest beta: " + beta);

        return beta;
    }

    /**
     * Get the latest alpha release.
     * @author Griefed
     * @return String. Returns the latest alpha release. If no alpha release is available, <code>no_alphas</code> is returned.
     * @throws NumberFormatException Thrown if a versions can not be parsed into integers.
     */
    protected String latestAlpha() throws NumberFormatException {

        List<String> alphaVersions = allAlphaVersions();
        String alpha = "no_alphas";

        if (alphaVersions != null) {

            alpha = alphaVersions.get(0);

            for (String alphaVersion : alphaVersions) {

                if (isNewOrSameSemanticVersion(alpha, alphaVersion) && Integer.parseInt(alphaVersion.substring(12)) > Integer.parseInt(alpha.substring(12))) {
                    alpha = alphaVersion;
                }
            }

        }

        LOG.debug("Latest alpha: " + alpha);

        return alpha;
    }

    public abstract void refresh() throws HttpClientErrorException, JsonProcessingException ;

    protected void setAllVersions() {
        this.allVersions = allVersions();
    }

    protected List<String> getAllVersions() {
        return allVersions;
    }

    protected abstract String latestVersion(boolean checkForPreRelease);

    protected abstract String getDownloadUrl(String version);

    protected abstract void setRepository() throws JsonProcessingException, HttpClientErrorException;

    public abstract List<String> getAssetsDownloadUrls(String version);

    protected ObjectMapper getObjectMapper() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
        objectMapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
        return objectMapper;
    }

}

File: serverpackcreator-app/src/main/kotlin/de/griefed/serverpackcreator/app/updater/versionchecker/VersionChecker.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Griefed
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package de.griefed.serverpackcreator.app.updater.versionchecker

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import de.griefed.serverpackcreator.api.utilities.common.Comparison
import de.griefed.serverpackcreator.api.utilities.common.SemanticVersionComparator
import org.apache.logging.log4j.kotlin.cachedLoggerOf
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.*

/**
 * Baseclass from which GitHub and GitLab checks extend. This class mainly provides the logic for comparing versions against
 * each other to find out which is newer. Extend from this class if you want to implement your own checkers, for platforms
 * like Gitea or anything else.
 * @author Griefed
 */
@Suppress("unused")
abstract class VersionChecker {
    private val log by lazy { cachedLoggerOf(this.javaClass) }
    protected var allVersions: List<String>? = null
        private set

    /**
     * Check whether an update/newer version is available for the given version. If you want to check for PreReleases, too,
     * then make sure to pass `true` for `checkForPreReleases`.
     * @author Griefed
     * @param currentVersion The current version of the app.
     * @param checkForPreReleases `false` if you do not want to check for PreReleases. `true`
     * if you want to check for PreReleases as well.
     * @return Returns a concatenated String whether an update is available or not. Examples:<br></br>
     * No update available: `No updates available.`<br></br>
     * New release available: `Current version: 2.0.0. A new release is available: 2.1.1. Download available at: https://github.com/Griefed/ServerPackCreator/releases/tag/2.1.1`
     * New prerelease available: `Current version: 2.0.0. A new PreRelease is available: 3.0.0-alpha.14. Download available at: https://github.com/Griefed/ServerPackCreator/releases/tag/3.0.0-alpha.14`
     */
    fun checkForUpdate(currentVersion: String, checkForPreReleases: Boolean): String {
        log.debug("Current version: $currentVersion")
        return try {
            val newVersion = isUpdateAvailable(currentVersion, checkForPreReleases)
            if (newVersion == "up_to_date") {
                "No updates available."
            } else {
                newVersion + ";" + getDownloadUrl(newVersion!!)
            }
        } catch (ex: NumberFormatException) {
            log.error("A version could not be parsed into integers.", ex)
            "No updates available."
        }
    }

    /**
     * Check for new versions in beta, alpha and regular release channels. If `checkForPreRelease` is false,
     * only regular releases are checked.
     * @author Griefed
     * @param currentVersion The current version of the app.
     * @param checkForPreReleases `false` if you do not want to check for PreReleases. `true`
     * if you want to check for PreReleases as well.
     * @return Returns the available update version. If no update is available, then `up_to_date` is returned.
     */
    protected fun isUpdateAvailable(currentVersion: String, checkForPreReleases: Boolean): String? {
        if (checkForPreReleases) {
            if (isNewBetaAvailable(currentVersion)) {
                return latestBeta()
            }
            if (isNewAlphaAvailable(currentVersion)) {
                return latestAlpha()
            }
        }
        if (SemanticVersionComparator.compareSemantics(currentVersion, latestVersion(checkForPreReleases)!!, Comparison.NEW)) {
            return latestVersion(checkForPreReleases)
        } else if (currentVersion.matches("\\d+\\.\\d+\\.\\d+-(alpha|beta)\\.\\d+".toRegex()) &&
            SemanticVersionComparator.compareSemantics(currentVersion, latestVersion(false)!!, Comparison.EQUAL)
        ) {
            return latestVersion(false)
        }
        return "up_to_date"
    }

    /**
     * Check for a new alpha version.
     * @author Griefed
     * @param currentVersion The current version to check against available alpha versions.
     * @return Returns true if a new alpha release is found.
     * @throws NumberFormatException Thrown if the passed `currentVersion` can not be
     * parsed into integers.
     */
    @Throws(NumberFormatException::class)
    private fun isNewAlphaAvailable(currentVersion: String): Boolean {

        // If no alpha releases are available, do not check for new alpha release.
        if (latestAlpha() == "no_alphas") {
            return false
        }
        val latestAlpha = latestAlpha()

        // A beta is never offered an alpha of the same version. That used to need an explicit guard
        // here; the channel ordering in isPreReleaseNewer now rules it out on its own.

        // Check if the given version is older than the latest alpha version by checking semantically. (1.2.3, 2.3.4, 6.6.6)
        return SemanticVersionComparator.compareSemantics(currentVersion, latestAlpha, Comparison.NEW)
                // Or, at the same version, if a new alpha — say alpha.5 for the given alpha.1 — is available.
                || (SemanticVersionComparator.compareSemantics(currentVersion, latestAlpha, Comparison.EQUAL_OR_NEW)
                && currentVersion.contains("-")
                && isPreReleaseNewer(currentVersion, latestAlpha))
    }

    /**
     * Check for a new beta version.
     * @author Griefed
     * @param currentVersion The current version to check against available beta versions.
     * @return Returns true if a new beta release is found.
     * @throws NumberFormatException Thrown if the passed `currentVersion` can not be
     * parsed into integers.
     */
    @Throws(NumberFormatException::class)
    private fun isNewBetaAvailable(currentVersion: String): Boolean {

        // If no beta releases are available, do not check for new beta release.
        if (latestBeta() == "no_betas") {
            return false
        }
        val latestBeta = latestBeta()

        // Check if the given version is older than the latest beta version by checking semantically. (1.2.3, 2.3.4, 6.6.6)
        return SemanticVersionComparator.compareSemantics(currentVersion, latestBeta, Comparison.NEW)
                // Or, at the same version, if a new beta — say beta.5 for the given beta.1 — is available.
                || (SemanticVersionComparator.compareSemantics(currentVersion, latestBeta, Comparison.EQUAL_OR_NEW)
                && currentVersion.contains("-")
                && isPreReleaseNewer(currentVersion, latestBeta))
    }

    /**
     * Check whether [newVersion] is a newer pre-release than [currentVersion], for two versions that
     * are already known to share the same semantic version.
     *
     * Channel first, number only as the tie-break: a beta supersedes an alpha of the same version no
     * matter what the two numbers are. Comparing the numbers alone made the answer depend on a
     * numeric accident — `alpha.5` would not be offered `beta.3`, because `3 > 5` is false.
     *
     * @author Griefed
     * @param currentVersion The current version for which we want to check for newer versions availability.
     * @param newVersion The new version with which we want to check if it is indeed newer than the current version.
     * @return True if the new version is a newer pre-release.
     */
    private fun isPreReleaseNewer(currentVersion: String, newVersion: String): Boolean {
        val channelDifference = preReleaseChannel(newVersion) - preReleaseChannel(currentVersion)
        return if (channelDifference != 0) {
            channelDifference > 0
        } else {
            preReleaseNumber(newVersion) > preReleaseNumber(currentVersion)
        }
    }

    /**
     * Rank of [version]'s pre-release channel, so the channels can be ordered: alpha, then beta, then
     * a plain release, which supersedes every pre-release of the same version.
     *
     * @author Griefed
     * @param version The version whose channel to rank.
     * @return `0` for an alpha, `1` for a beta, `2` for anything else.
     */
    private fun preReleaseChannel(version: String) = when {
        version.contains("alpha") -> 0
        version.contains("beta") -> 1
        else -> 2
    }

    /**
     * The number a pre-release counts up, e.g. `5` for `3.1.0-alpha.5`.
     *
     * A version carrying no pre-release suffix, or one whose suffix holds no number, yields `0` — it
     * is only ever compared within one channel by [isPreReleaseNewer], where a plain release already
     * outranks anything numbered.
     *
     * @author Griefed
     * @param version The version whose pre-release number to read.
     * @return The pre-release number, or `0` when the version carries none.
     */
    private fun preReleaseNumber(version: String) =
        version.substringAfter('-', "").substringAfter('.', "").toIntOrNull() ?: 0

    /**
     * Check whether [candidate] is newer than [current], comparing the semantic version first and
     * falling back to the pre-release ordering only when the two are the same version.
     *
     * This is what picking "the latest" of a channel needs. Requiring a candidate to be *both*
     * newer-or-equal **and** higher-numbered — as the latest-of-channel scans used to — drops a newer
     * version that restarted its count, so `3.2.0-beta.1` lost to `3.1.0-beta.3`.
     *
     * @author Griefed
     * @param current The version to beat.
     * @param candidate The version being considered.
     * @return True if the candidate supersedes the current one.
     */
    private fun isVersionNewer(current: String, candidate: String) = when {
        SemanticVersionComparator.compareSemantics(current, candidate, Comparison.NEW) -> true
        !SemanticVersionComparator.compareSemantics(current, candidate, Comparison.EQUAL) -> false
        else -> isPreReleaseNewer(current, candidate)
    }

    /**
     * Gather all beta versions in a list.
     * @author Griefed
     * @return Returns a list of all available beta versions. If no beta releases are available for the given
     * repository, `null` is returned.
     */
    private fun allBetaVersions(): List<String>? {
        val betaVersions: MutableList<String> = ArrayList(1000)

        // Check all available versions whether they contain beta and add them to a list of beta versions.
        if (allVersions != null) {
            for (version in allVersions!!) {
                if (version.contains("beta") && !betaVersions.contains(version)) {
                    betaVersions.add(version)
                }
            }
        }
        return if (betaVersions.isEmpty()) {
            null
        } else {
            betaVersions
        }
    }

    /**
     * Gather all alpha versions in a list.
     * @author Griefed
     * @return Returns a list of all available alpha versions. If no alpha releases are available for the given
     * repository, `null` is returned.
     */
    private fun allAlphaVersions(): List<String>? {
        val alphaVersions: MutableList<String> = ArrayList(1000)

        // Check all available versions whether they contain alpha and add them to a list of alpha versions.
        if (allVersions != null) {
            for (version in allVersions!!) {
                if (version.contains("alpha") && !alphaVersions.contains(version)) {
                    alphaVersions.add(version)
                }
            }
        }
        return if (alphaVersions.isEmpty()) {
            null
        } else {
            alphaVersions
        }
    }

    /**
     * Get the latest beta release.
     * @author Griefed
     * @return Returns the latest beta release. If no beta release is available, `no_betas` is returned.
     * @throws NumberFormatException Thrown if a version can not be parsed into integers.
     */
    @Throws(NumberFormatException::class)
    protected fun latestBeta(): String {
        val betaVersions = allBetaVersions()
        var beta = "no_betas"
        if (betaVersions != null) {
            beta = betaVersions[0]
            for (betaVersion in betaVersions) {
                if (isVersionNewer(beta, betaVersion)) {
                    beta = betaVersion
                }
            }
        }
        log.debug("Latest beta: $beta")
        return beta
    }

    /**
     * Get the latest alpha release.
     * @author Griefed
     * @return Returns the latest alpha release. If no alpha release is available, `no_alphas` is returned.
     * @throws NumberFormatException Thrown if a versions can not be parsed into integers.
     */
    @Throws(NumberFormatException::class)
    protected fun latestAlpha(): String {
        val alphaVersions = allAlphaVersions()
        var alpha = "no_alphas"
        if (alphaVersions != null) {
            alpha = alphaVersions[0]
            for (alphaVersion in alphaVersions) {
                if (isVersionNewer(alpha, alphaVersion)) {
                    alpha = alphaVersion
                }
            }
        }
        log.debug("Latest alpha: $alpha")
        return alpha
    }

    /**
     * Acquire the response from a given URL.
     * @author Griefed
     * @param requestUrl The URL to get the response from.
     * @return The response from the given URL.
     * @throws IOException Thrown if the requested URL can not be reached or if any other error occurs during the request.
     */
    @Throws(IOException::class)
    protected fun getResponse(requestUrl: URL): String {
        val httpURLConnection = requestUrl.openConnection() as HttpURLConnection
        httpURLConnection.requestMethod = "GET"
        if (httpURLConnection.responseCode != 200) throw IOException("Request for " + requestUrl + " responded with " + httpURLConnection.responseCode)
        val bufferedReader = BufferedReader(
            InputStreamReader(httpURLConnection.inputStream)
        )
        var inputLine: String?
        val response = StringBuilder()
        while (bufferedReader.readLine().also { inputLine = it } != null) {
            response.append(inputLine)
        }
        bufferedReader.close()
        return response.toString()
    }

    protected val objectMapper: ObjectMapper
        get() {
            val objectMapper = ObjectMapper()
            objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
            objectMapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
            return objectMapper
        }

    protected abstract fun allVersions(): List<String>?
    /**
     * Re-read the repository's version list, so subsequent checks see releases published since the
     * last call.
     *
     * @author Griefed
     * @return This checker, for chaining.
     * @throws IOException if the repository could not be reached.
     */
    @Throws(IOException::class)
    abstract fun refresh(): VersionChecker
    protected fun setAllVersions() {
        allVersions = allVersions()
    }

    protected abstract fun latestVersion(checkForPreRelease: Boolean): String?
    protected abstract fun getDownloadUrl(version: String): String
    @Throws(IOException::class)
    protected abstract fun setRepository()
    /**
     * The available update for [currentVersion] as a repository-specific [Update], carrying the
     * release's own metadata (description, assets) rather than just a version string.
     *
     * @author Griefed
     * @param currentVersion The current version of the app.
     * @param checkForPreReleases Whether alpha and beta releases count as updates.
     * @return The update, or empty when the current version is the newest.
     */
    abstract fun check(currentVersion: String, checkForPreReleases: Boolean): Optional<Update>
}